Refactor agent functionality

Yukinobu-Mine opened the pull request:
*Issue #, if available:*

*Description of changes:*
This refactoring involved the following steps:

Step 1: Preparation https://github.com/aws-samples/bedrock-claude-chat/commit/d658b9521d5dee2a7093a86025b03433835ec243
- Introduce boto3_stubs (and remove similar custom definitions in `bedrock.py`)
- Improve type safety
- Remove `AgentContentModel`s and integrated into `ContentModel`
- Simplify code to fetch `MessageModel`

Step 2: Change the type of tool result to list. https://github.com/aws-samples/bedrock-claude-chat/commit/2b0bc843a049867266df9baa12bdd3d0c64c0fa7
- Backend
  - `RunResult.body`
  - `ConverseApiToolResult.content`
  - `ToolResultContentModelBody.content`
  - `ToolResultContentBody.content`
  - Return value of `internet_search()`

- Frontend
  - `AgentToolResult.content`
  - `ToolCardProps.content`
  - `AgentToolsProps.content`
  - `AgentEvent.content`

Step 3: Unify the duplicated implementation of chat between WebSocket API and REST API. https://github.com/aws-samples/bedrock-claude-chat/commit/8294c259dabb0f7e34385ec52d753b8dd6c20404
- Migrate specific implementation in `websocket.process_chat_input()` to `usecases.chat()`.
- Remove the implementation of chat in `websocket.process_chat_input()`, and use `usecases.chat()` instead.

Step 4: Unify the implementation of agent chat and normal chat. https://github.com/aws-samples/bedrock-claude-chat/commit/a3fe6ca972504a881d85e22d7ed132af754e6f2c
- Add agent (tool use) functionality to `ConverseApiStreamHandler`.
  - Add toolConfig support to `bedrock.compose_args_for_converse_api()`.
- Migrate the functionality of `AgentRunner` to `usecases.chat()`.
  - Migrate the test cases of agent functionality to `test_chat.py`.
  - Remove `agent.py` and `test_agent.py`.
- Change the parameters of `AgentTool.run()`.
  - Add `tool_use_id`.
  - Change `arg: T` to `input: dict[str, JsonValue]`.
- Now agent chat supports multi-modal content.
  - Use `InputChatContent` in `ChatPage` of frontend even if `bot.hasAgent`.

Step 5: Implement 'Retrieved Context Citation' of agent chat. https://github.com/aws-samples/bedrock-claude-chat/commit/6aa454b523a755b2140f9b2b00641b1a08fe1d00
- Redefine the schema of `RelatedDocument`.
  - Add database entity `RELATED_DOCUMENT`.
  - Store `SearchResult` and `ToolResult` as related document.
  - Delete related documents when a conversation is deleted.
- Add REST API `GET /conversation/{conversation_id}/related-documents` to retrieve related documents of a conversation.
  - Remove REST API `POST /conversation/related-documents`.
  - Now, related documents are linked to the entire conversation, not just a message.
- Refactor `ChatMessageMarkdown`
  - Convert citations with `[^source_id]` to sequential numbers in order of appearance such as `[^1]`.
  - Separate the component that displays a related document as `RelatedDocumentViewer`.
  - Instead of passing `ChatMessageMarkdown` the related document itself, pass a callback to retrieve the related document.
    - Treat used chunks as related documents as well.
  - Disable links to citation sources while streaming.
  - Optimize handling of streaming cursors.
- Improvement of internet search tool:
  - Treat search results as citable related documents, and display page titles and links in `RelatedDocumentViewer`.

Step 6: Change the specifications for `AGENT_TOOL_RESULT` notification. https://github.com/aws-samples/bedrock-claude-chat/commit/d12ecc695153d8eddba5b363f8144f9840b8e013
- Specification changes
  - Split `AGENT_TOOL_RESULT` to `AGENT_TOOL_RESULT` and `AGENT_RELATED_DOCUMENT`
    - `AGENT_TOOL_RESULT` contains only `status` of tool result.
    - `AGENT_RELATED_DOCUMENT` contains `RelatedDocument` of tool result.
  - Change `ChatMessage`'s property `getRelatedDocument` back to `relatedDocuments`.
    - Also changed `ChatMessageMarkdown` and `AgentToolList`.
  - Add `relatedDocuments` to `ToolCard`.
    - Show tool result as `RelatedDocument` instead of `AgentToolResultContent`.
    - For backward compatibility, show `AgentToolResultContent` if `RelatedDocument` is not found.

- Type name changes
  - `ConverseApiToolResult` and `RunResult`-> Integrated to `ToolRunResult`
    - Changed result type: `ToolResultModel` -> `RelatedDocumentModel`
    - Move the conversion code from `usecases/chat.py` to `agents/tools/agent_tool.py`
  - `AgentResultType` -> `ToolFunctionResult`
  - `AgentMessageModel` -> `SimpleMessageModel`
  - `AgentMessage` -> `SimpleMessage`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Yukinobu-Mine committed:
Refactor agent implementation: Phase 1
- Introduce boto3_stubs
- Improve type safety
- Remove `AgentContentModel`s and integrated into `ContentModel`
- Simplify code to fetch `MessageModel`

Yukinobu-Mine committed:
Change the type of tool result to list.
Backend
- `RunResult.body`
- `ConverseApiToolResult.content`
- `ToolResultContentModelBody.content`
- `ToolResultContentBody.content`
- Return value of `internet_search()`

Frontend
- `AgentToolResult.content`
- `ToolCardProps.content`
- `AgentToolsProps.content`
- `AgentEvent.content`

Yukinobu-Mine committed:
Unify the duplicated implementation of chat between WebSocket API and REST API.
- Migrate specific implementation in `websocket.process_chat_input()` to `usecases.chat()`.
- Remove the implementation of chat in `websocket.process_chat_input()`, and use `usecases.chat()` instead.

Yukinobu-Mine committed:
Unify the implementation of agent chat and normal chat.
- Add agent (tool use) functionality to `ConverseApiStreamHandler`.
  - Add `toolConfig` support to `bedrock.compose_args_for_converse_api()`.
- Migrate the functionality of `AgentRunner` to `usecases.chat()`.
  - Migrate the test cases of agent functionality to `test_chat.py`.
  - Remove `agent.py` and `test_agent.py`.
- Change the parameters of `AgentTool.run()`.
  - Add `tool_use_id`.
  - Change `arg: T` to `input: dict[str, JsonValue]`.
- Now agent chat supports multi-modal content.
  - Use `InputChatContent` even if `bot.hasAgent`.

Yukinobu-Mine committed:
Update mypy
- mypy 1.11.2 -> 1.13.0

Yukinobu-Mine committed:
Fix mypy errors

Yukinobu-Mine committed:
Fix python format warnings

Yukinobu-Mine committed:
Fix errors of test cases.

statefb committed:
fix: correct content parsing logic in ToolCard to prevent duplicate entries

- Replaced flatMap with map to correctly handle each content item individually
- Removed dependency on content[0] to avoid repeated display of the first item

statefb committed:
chore: remove duplicated test for agent

statefb reviewed (approved):
Incredible job on this refactoring!

Yukinobu-Mine committed:
Add agent thought to `AgentToolList`.
- Improve type safety of frontend's `Content` and `AgentContent`.
- Support agent thought and multi-turn tool use in `agentThinkingState`.
- Add agent thought and localized tool name to `AgentToolList`.
  - Add localized name of `knowledge_base_tool`.
- Display text content as a thought that received before `AGENT_THINKING` message.

Yukinobu-Mine committed:
Fix: Error handling in `ConverseApiStreamHandler`

Yukinobu-Mine committed:
Remove `output` from result of `knowledge_base_tool`.

Yukinobu-Mine committed:
Fix: Description of `ContentModel.body`

statefb commented:
Multi turn tool use example:
![image](https://github.com/user-attachments/assets/03458e78-96dd-47b8-859f-b37f726dd61a)

Yukinobu-Mine committed:
Implement 'Retrieved Context Citation' of agent chat.
- Redefine the schema of related documents.
- Add database entity `RELATED_DOCUMENT`.
  - Store `SearchResult` and `ToolResult` as related document.
  - Delete related documents when a conversation is deleted.
- Add REST API `GET /conversation/{conversation_id}/related-documents` to retrieve related documents of a conversation.
  - Remove REST API `POST /conversation/related-documents`.
  - Now, related documents are linked to the entire conversation, not just a message.
- Refactor `ChatMessageMarkdown`
  - Convert citations with '[^source_id]' to sequential numbers in order of appearance such as '[^1]'.
  - Separate the component that displays a related document as `RelatedDocumentViewer`.
  - Instead of passing `ChatMessageMarkdown` the related document itself, pass a callback to retrieve the related document.
    - Treat used chunks as related documents as well.
  - Disable links to citation sources while streaming.
  - Optimize handling of streaming cursors.
- Improvement of internet search tool:
  - Treat search results as citable related documents, and display page titles and links in `RelatedDocumentViewer`.

Yukinobu-Mine committed:
Omit `content` of `AGENT_TOOL_RESULT` notification if the content size exceeds 128KB
- If the result of agent has `source_id`, use it as the `source_id` of `RelatedDocument`.

Yukinobu-Mine committed:
Fix: When notification via WebSocket fails, the entire process is treated as failure.

Yukinobu-Mine committed:
Merge branch 'v2' into refactor-agent

Yukinobu-Mine committed:
Change the specifications for `AGENT_TOOL_RESULT` notification.
Specification changes
- Split `AGENT_TOOL_RESULT` to `AGENT_TOOL_RESULT` and `AGENT_RELATED_DOCUMENT`
  - `AGENT_TOOL_RESULT` contains only `status` of tool result.
  - `AGENT_RELATED_DOCUMENT` contains `RelatedDocument` of tool result.
- Change `ChatMessage`'s property `getRelatedDocument` back to `relatedDocuments`.
  - Also changed `ChatMessageMarkdown`, `AgentToolList`.
- Add `relatedDocuments` to `ToolCard`.
  - Show tool result as `RelatedDocument` instead of `AgentToolResultContent`.
  - For backward compatibility, show `AgentToolResultContent` if `RelatedDocument` is not found.

Type name changes
- `ConverseApiToolResult` and `RunResult`-> Integrated to `ToolRunResult`
  - Changed result type: `ToolResultModel` -> `RelatedDocumentModel`
  - Move the conversion code from `usecases/chat.py` to `agents/tools/agent_tool.py`
- `AgentResultType` -> `ToolFunctionResult`
- `AgentMessageModel` -> `SimpleMessageModel`
- `AgentMessage` -> `SimpleMessage`

Yukinobu-Mine committed:
Display search results of Knowledge Base even when agents are not enabled.
